indexer/tasks/src/multiera/multiera_asset_utxo.rs

```diff
@@ -0,0 +1,207 @@
+use crate::multiera::dex::common::filter_outputs_and_datums_by_address;
+use crate::multiera::multiera_txs::MultieraTransactionTask;
+use crate::multiera::multiera_used_inputs::MultieraUsedInputTask;
+use crate::multiera::multiera_used_outputs::MultieraOutputTask;
+use crate::types::AddressCredentialRelationValue;
+use cardano_multiplatform_lib::error::DeserializeError;
+use cml_core::serialization::{FromBytes, ToBytes};
+use cml_crypto::RawBytesEncoding;
+use entity::sea_orm::Condition;
+use entity::transaction_output::Model;
+use entity::{
+    prelude::*,
+    sea_orm::{prelude::*, DatabaseTransaction, Set},
+};
+use pallas::ledger::primitives::babbage::DatumOption;
+use pallas::ledger::primitives::Fragment;
+use pallas::ledger::traverse::{Asset, MultiEraInput, MultiEraOutput};
+use projected_nft_sdk::{Owner, State, Status};
+use sea_orm::{FromQueryResult, JoinType, QuerySelect, QueryTrait};
+use std::collections::{BTreeSet, HashMap};
+
+carp_task! {
+  name MultieraAssetUtxos;
+  configuration EmptyConfig;
+  doc "Parses utxo movements for native assets";
+  era multiera;
+  dependencies [MultieraUsedInputTask, MultieraOutputTask];
+  read [multiera_txs, multiera_outputs, multiera_used_inputs_to_outputs_map];
+  write [];
+  should_add_task |block, _properties| {
+    !block.1.is_empty()
+  };
+  execute |previous_data, task| handle(
+      task.db_tx,
+      task.block,
+      &previous_data.multiera_txs,
+      &previous_data.multiera_outputs,
+      &previous_data.multiera_used_inputs_to_outputs_map,
+  );
+  merge_result |previous_data, _result| {
+  };
+}
+
+async fn handle(
+    db_tx: &DatabaseTransaction,
+    block: BlockInfo<'_, MultiEraBlock<'_>, BlockGlobalInfo>,
+    multiera_txs: &[TransactionModel],
+    multiera_outputs: &[TransactionOutputModel],
+    multiera_used_inputs_to_outputs_map: &BTreeMap<Vec<u8>, BTreeMap<i64, OutputWithTxData>>,
+) -> Result<(), DbErr> {
+    let mut queued_inserts = vec![];
+
+    // this stores the result before searching for the asset ids in the table
+    struct PartialEntry {
+        utxo_id: i64,
+        amount: Option<i64>,
+        tx_id: i64,
+        // policy_id + asset_name
+        asset: (Vec<u8>, Vec<u8>),
+    }
+
+    let mut condition = Condition::any();
+
+    let outputs_map: HashMap<_, _> = multiera_outputs
+        .iter()
+        .map(|output| ((output.tx_id, output.output_index), output))
+        .collect();
+
+    for (tx_body, cardano_transaction) in block.1.txs().iter().zip(multiera_txs) {
+        for input in tx_body.inputs().iter().chain(tx_body.collateral().iter()) {
+            let utxo = multiera_used_inputs_to_outputs_map
+                .get(input.hash().as_ref())
+                .and_then(|by_index| by_index.get(&(input.index() as i64)));
+
+            let utxo = if let Some(utxo) = utxo {
+                utxo
+            } else {
+                // this can happen if the tx was not valid, in which case the
+                // input is not spent.
```

Comment:
I think this comment is wrong because in the case of a failed tx, the collateral input is consumed
I think a more accurate comment would be something like this
```
We iterate over the combination of <inputs, collateral>
If tx succeeded, inputs are consumed and we skip collateral
If tx failed, we skip inputs and consume collateral
```